python_modules/dagster/dagster/_core/definitions/decorators/decorator_assets_definition_builder.py

```diff
@@ -380,19 +380,22 @@ def from_multi_asset_specs(
                     upstream_deps[dep.asset_key] = dep
 
         # get which asset keys have inputs set
-        loaded_upstreams = build_and_validate_named_ins(fn, asset_in_map, deps=set())
-        unexpected_upstreams = {key for key in loaded_upstreams.keys() if key not in upstream_deps}
-        if unexpected_upstreams:
-            raise DagsterInvalidDefinitionError(
-                f"Asset inputs {unexpected_upstreams} do not have dependencies on the passed"
-                " AssetSpec(s). Set the deps on the appropriate AssetSpec(s)."
-            )
-        remaining_upstream_deps = [
-            dep for key, dep in upstream_deps.items() if key not in loaded_upstreams
-        ]
         named_ins_by_asset_key = build_and_validate_named_ins(
-            fn, asset_in_map, deps=remaining_upstream_deps
+            fn, asset_in_map, deps=upstream_deps.values()
         )
+        # We expect that asset_ins are a subset of asset_deps. The reason we do not check this in
+        # `build_and_validate_named_ins` is because in other decorator pathways, we allow for argument-based
+        # dependencies which are not specified in deps (such as the asset decorator).
+        asset_in_keys = set(named_ins_by_asset_key.keys())
+        if asset_in_keys - set(upstream_deps.keys()):
+            input_names_not_in_deps = [
+                named_ins_by_asset_key[key].input_name
+                for key in asset_in_keys - set(upstream_deps.keys())
+            ]
+            raise DagsterInvalidDefinitionError(
+                f"Asset inputs {input_names_not_in_deps} specified as asset inputs, but are not specified as dependencies on any constituent asset specs."
```

Comment:
same comment as https://github.com/dagster-io/dagster/pull/26465/files#r1884541448 - i think hitting this exception could be pretty confusing (especially for a new user dealing w these concepts for the first time). Giving the specific class names would be a big improvement